Allow adm group to read logs

patrobinson opened the pull request:
Traditionally the `adm` group is allowed to read system logs, but not write to them.

This enables log agents that run as a non-root user to read the logs. We want the audit logs to be readable by log agents but probably not the session logs.

Commit:
Allow adm group to read logs

Traditionall the adm group is allowed to read system logs, but not write to them.

This enables log agents that run as a non-root user to read the logs. We want the audit logs to be readable by log agents but probably not the session logs

gravitational-jenkins commented:
Can one of the admins verify this patch?

Commit:
Merge branch 'master' into allow-non-root-read-log

Commit:
Also ensure the log directory is listable by those in the adm group

kontsevoy reviewed (approved):
looks good.

klizhentas commented on lib/events/auditlog.go:
```diff
@@ -97,10 +103,17 @@ type AuditLog struct {
 func NewAuditLog(dataDir string, recordSessions bool) (IAuditLog, error) {
+	// create the log directory first, with slightly broader permissions
+	if err := os.MkdirAll(dataDir, 0770); err != nil {
+		return nil, trace.Wrap(err)
```
please use `trace.ConvertSystemError` to convert access denied erorrs

Commit:
Convert system error into more readable error

patrobinson commented:
Note, the tests need to be run as root now because we're trying to modify a directory that was originally run as root. Is it worth modifying the code to make the root uid configurable and making it the user who is running the tests?

klizhentas commented:
@patrobinson this is actually a very good point:

* Teleport tests are first-class citizen, so we should probably pass the UID/GID as parameters so tests can pick current UIDs

* Sometimes teleport can be run as non-root and we should not crash

Taking those two points into consideration, we should:

* Based on the current UID in lib/service/service.go, set a group to `adm` group in case if we run as root, and to current user otherwise.

* Pass both UID/GID parameter to AuditLog when initializing it

* Tests should  default to current UID when running

Commit:
Ensure we are part of the adm group

If we're not, trying to change the group of the log dir will fail with 'operation not permitted'

patrobinson commented:
@klizhentas I stuck this logic in AuditLog instead of Service because:
- It seemed like that logic belongs there because only AuditLog cares about it
- It means I don't have to change all the tests that call `NewAuditLog()`

Let me know if this works for you.

kontsevoy commented:
@patrobinson @klizhentas see `man fakeroot`. This exists _exactly_ for this use case.

kontsevoy commented:
I would say that something like a simple integration test running from under fakeroot can cover this.

patrobinson commented:
@kontsevoy it's a little difficult to implement without duplicating all the integration test suite, since I can only apply build tags per file, to selectively run some tests with fakeroot and others without.

I can just run the integration tests both with and without `fakeroot`, but if you're already root when you run them they aren't an accurate test. I also have no visibility into your CI environment to understand if it's running these integration tests and if it runs as a root user.

klizhentas commented:
@patrobinson we have discussed your PR yesterday, here are my notes:

* I would like to pass UID and GID as a backend parameters in configuration file so folks can set them to whatever make sense
* Your defaults 0/4 uid/gid make sense, so I would use them
* If the directory already exists, I would not change the permissions as the migration could fail
* Tests will use current UID/GID to avoid conflicts

That increases the scope of the PR, so I will base off your PR and will factor in your changes if you don't mind.

patrobinson commented:
@klizhentas happy for you to take this over if you like, thanks

kontsevoy commented:
@klizhentas lets keep adm GUID (only for brand-new `../log` dirs) as hard-coded for now. Seems like a safe change to make. Adding a proper audit log configuration section makes this change more time-consuming.

klizhentas commented:
@patrobinson I have implemented the same logic in a slightly different way, so closing your PR in favor of this one that we are going to merge:

https://github.com/gravitational/teleport/pull/1463

Thanks for your idea and PR, this is very helpful!
